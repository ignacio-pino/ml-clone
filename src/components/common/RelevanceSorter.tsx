import { HStack, Select, Text } from "@chakra-ui/react";
import { createSearchParams, useLocation, useNavigate } from "react-router-dom";
import { ISortData } from "../../Interfaces";

export default function RelevanceSorter({
  selectedSort,
  availableSorts,
}: {
  selectedSort: ISortData;
  availableSorts: ISortData[];
}) {
  const { search } = useLocation();
  const searchParams = new URLSearchParams(search);

  const navigate = useNavigate();

  let params = Object.fromEntries(searchParams.entries());

  const handleChange = (event: React.FormEvent<HTMLSelectElement>) => {
    const target = event.target as HTMLSelectElement;
    params = { ...params, sort: target.value };
    navigate({
      pathname: "/products",
      search: `${createSearchParams(params)}`,
    });
  };

  return (
    <HStack alignSelf="flex-end" my={"1rem"} w={"20rem"}>
      {" "}
      <Text as="b">Ordenar por</Text>{" "}
      <Select
        placeholder={selectedSort.name}
        w={"10rem"}
        value={selectedSort.id}
        onChange={handleChange}
      >
        {availableSorts.map(({ id, name }) => (
          <option value={id}>{name}</option>
        ))}
      </Select>
    </HStack>
  );
}
