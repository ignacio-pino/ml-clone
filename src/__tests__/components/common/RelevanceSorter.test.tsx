import { render, screen, fireEvent } from "@testing-library/react";
import RelevanceSorter from "../../../components/common/RelevanceSorter";
import { MemoryRouter, createSearchParams } from "react-router-dom";

const mockedUseNavigate = vitest.fn();

vitest.mock("react-router-dom", () => ({
  ...require("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
  useLocation: () => ({ search: "query=test" }),
}));

describe("RelevanceSorter Component", () => {
  const mockAvailableSorts = [
    { id: "price_asc", name: "Price: Low to High" },
    { id: "price_desc", name: "Price: High to Low" },
  ];
  const mockSelectedSort = { id: "relevance", name: "Relevance" };

  beforeEach(() => {
    mockedUseNavigate.mockClear();
  });

  it("matches snapshot", () => {
    const params = { sort: "relevance" };
    const container = render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <RelevanceSorter
          selectedSort={mockSelectedSort}
          availableSorts={mockAvailableSorts}
        />
      </MemoryRouter>
    ).container;

    expect(container).toMatchSnapshot();
  });

  it("renders select with correct options", () => {
    const params = { sort: "relevance" };
    render(
      <MemoryRouter
        initialEntries={[
          { pathname: "/products", search: `${createSearchParams(params)}` },
        ]}
      >
        <RelevanceSorter
          selectedSort={mockSelectedSort}
          availableSorts={mockAvailableSorts}
        />
      </MemoryRouter>
    );

    const selectElement = screen.getByRole("combobox");
    const optionElements = screen.getAllByRole("option");

    expect(selectElement).toHaveTextContent(mockSelectedSort.name);
    expect(optionElements).toHaveLength(mockAvailableSorts.length + 1);
  });

  it("calls the handleChange function with correct params when an option is selected", () => {
    render(
      <MemoryRouter>
        <RelevanceSorter
          selectedSort={mockSelectedSort}
          availableSorts={mockAvailableSorts}
        />
      </MemoryRouter>
    );

    const selectElement = screen.getByRole("combobox");
    fireEvent.change(selectElement, { target: { value: "price_asc" } });

    expect(mockedUseNavigate).toHaveBeenCalledWith({
      pathname: "/products",
      search: "query=test&sort=price_asc",
    });
  });
});
