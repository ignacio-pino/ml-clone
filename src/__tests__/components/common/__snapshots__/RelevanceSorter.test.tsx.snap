// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RelevanceSorter Component > matches snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-9euhlz"
  >
    <b
      class="chakra-text css-0"
    >
      Ordenar por
    </b>
    <div
      class="chakra-select__wrapper css-xqwyme"
    >
      <select
        class="chakra-select css-1ik61og"
      >
        <option
          value=""
        >
          Relevance
        </option>
        <option
          value="price_asc"
        >
          Price: Low to High
        </option>
        <option
          value="price_desc"
        >
          Price: High to Low
        </option>
      </select>
      <div
        class="chakra-select__icon-wrapper css-162mkon"
      >
        <svg
          aria-hidden="true"
          class="chakra-select__icon"
          focusable="false"
          role="presentation"
          style="width: 1em; height: 1em; color: currentColor;"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
